libcudacxx/include/cuda/std/__type_traits/common_type.h

```diff
@@ -127,7 +127,44 @@ _LIBCUDACXX_INLINE_VAR constexpr bool __has_common_type = false;
 
 template <class _Tp, class _Up>
 _LIBCUDACXX_INLINE_VAR constexpr bool __has_common_type<_Tp, _Up, void_t<common_type_t<_Tp, _Up>>> = true;
-#endif
+#endif // _CCCL_STD_VER >= 2014
+
+#if defined(_LIBCUDACXX_HAS_NVFP16) && !defined(__CUDA_NO_HALF_CONVERSIONS__) && !defined(__CUDA_NO_HALF_OPERATORS__)
+template <>
+struct common_type<__half, __half>
+{
+  using type = __half;
+};
+template <class T>
+struct common_type<__half, T> : common_type<float, T>
+{};
+template <class T>
+struct common_type<T, __half> : common_type<T, float>
+{};
+#endif // _LIBCUDACXX_HAS_NVFP16 && !__CUDA_NO_HALF_CONVERSIONS__ && !__CUDA_NO_HALF_OPERATORS__
+
+#if defined(_LIBCUDACXX_HAS_NVBF16) && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__) \
+  && !defined(__CUDA_NO_BFLOAT16_OPERATORS__)
+template <>
+struct common_type<__nv_bfloat16, __nv_bfloat16>
+{
+  using type = __nv_bfloat16;
+};
+template <class T>
+struct common_type<__nv_bfloat16, T> : common_type<float, T>
+{};
+template <class T>
+struct common_type<T, __nv_bfloat16> : common_type<T, float>
+{};
+
+// __half and __nvbfloat16 have unordered conversion rank
+template <>
+struct common_type<__half, __nv_bfloat16>
+{};
+template <>
+struct common_type<__nv_bfloat16, __half>
+{};
```

Comment:
Question: Does `_LIBCUDACXX_HAS_NVBF16` imply that `__half` is available? If not, then you may need to guard this code also by `_LIBCUDACXX_HAS_NVFP16`.

Reply:
Yes, the nvbfloat16 header depends on the half header